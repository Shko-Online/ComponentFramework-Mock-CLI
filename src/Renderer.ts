import { IData } from 'ShkoOnline.types';
import { readFile } from 'fs/promises';
import Handlebars from 'handlebars';

export default class Renderer {
    constructor() {
        Handlebars.registerHelper('GetPropertyMocks', this.GetPropertyMocks.bind(this));
        Handlebars.registerHelper('ControlToPropertyMock', this.ControlToPropertyMock.bind(this));
        Handlebars.registerHelper('ControlToType', this.ControlToType.bind(this));
        Handlebars.registerHelper('isVirtualComponent', this.isVirtualComponent.bind(this));
    }

    private GetPropertyMocks(inputs: IData['Inputs']) {
        const propertyMocks: string[] = [];
        Object.getOwnPropertyNames(inputs).forEach(p => {
            if (!propertyMocks.some(pm => pm === inputs[p])) {
                propertyMocks.push(this.ControlToPropertyMock(inputs[p]));
            }
        });
        return propertyMocks.join(', ');
    }

    private isVirtualComponent(mockGenerator: IData['MockGenerator']) {
        return mockGenerator === 'ComponentFrameworkMockGeneratorReact';
    }

    private ControlToPropertyMock(control: IData['Inputs'][any]) {
        switch (control) {
            case 'DateAndTime.DateOnly':
                return 'DateTimePropertyMock';
            case 'Multiple':
                return 'StringPropertyMock';
            case 'SingleLine.Text':
                return 'StringPropertyMock';
            case 'TwoOptions':
                return 'TwoOptionsPropertyMock';
            default:
                return '';
        }
    }

    private ControlToType(control: IData['Inputs'][any]) {
        switch (control) {
            case 'DateAndTime.DateOnly':
                return 'Date';
            case 'Multiple':
            case 'SingleLine.Text':
                return 'string';
            case 'TwoOptions':
                return 'boolean';
            default:
                return '';
        }
    }

    async Render(data: IData) {
        const buffer = await readFile(require.resolve('./renderGenerator.template.hbs'));
        const template = Handlebars.compile(buffer.toString());


        const dts = template({ data });

        return dts;
    }
}