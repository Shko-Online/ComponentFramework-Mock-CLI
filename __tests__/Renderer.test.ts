import { describe, expect, test } from '@jest/globals';
import Renderer from '../src/Renderer';

describe('Renderer', () => {
    test('can load template', async () => {
        const renderer = new Renderer();
        var result = await renderer.Render({
            Component: 'Calendar',
            ComponentFolder: '../Calendar',
            ComponentIndex: 'index',
            Inputs: [
                { name: "Theme", type: "Multiple", usage: "input" },
                { name: "BackgroundColor", type: "SingleLine.Text", usage: "input" },
                { name: "Language", type: "SingleLine.Text", usage: 'input' },
                { name: "AccessibilityLabel", type: "SingleLine.Text", usage: 'input' },
                { name: "SelectedDateValue", type: "DateAndTime.DateOnly", usage: 'bound' },
                { name: "MinDate", type: "DateAndTime.DateOnly", usage: 'input' },
                { name: "MaxDate", type: "DateAndTime.DateOnly", usage: 'input' },
                { name: "ShowGoToToday", type: "TwoOptions", usage: 'input' },
                { name: "MonthPickerVisible", type: "TwoOptions", usage: 'input' },
                { name: "DayPickerVisible", type: "TwoOptions", usage: 'input' },
                { name: "HighlightSelectedMonth", type: "TwoOptions", usage: 'input' },
                { name: "HighlightCurrentMonth", type: "TwoOptions", usage: 'input' },
                { name: "ShowSixWeeksByDefault", type: "TwoOptions", usage: 'input' },
                { name: "ShowWeekNumbers", type: "TwoOptions", usage: 'input' },
                { name: "InputEvent", type: "SingleLine.Text", usage: 'input' },
                {
                    name: "FirstDayOfWeek", type: "Enum", usage: 'input', enumValues: [
                        { name: "Sunday", value: "Sunday" },
                        { name: "Monday", value: "Monday" },
                        { name: "Tuesday", value: "Tuesday" },
                        { name: "Wednesday", value: "Wednesday" },
                        { name: "Thursday", value: "Thursday" },
                        { name: "Friday", value: "Friday" },
                        { name: "Saturday", value: "Saturday" }
                    ]
                }
            ],
            MockGenerator: 'ComponentFrameworkMockGeneratorReact'
        });

        expect(result).toMatchSnapshot();
    });
});