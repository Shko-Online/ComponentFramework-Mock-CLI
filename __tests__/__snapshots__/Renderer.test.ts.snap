// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderer can load template 1`] = `
"/* 
    This file was generated with @shko.online/componentframework-mock-cli (pcf-sb)
    Do not manually update this file as it gets overwritten by pcf-sb based on the Manifest configuration.
*/

import type { Meta, StoryObj } from '@storybook/html';

import type { IInputs, IOutputs } from '../Calendar/generated/ManifestTypes';

import { useArgs, useEffect } from "@storybook/preview-api";

import { ComponentFrameworkMockGeneratorReact, StringPropertyMock, DateTimePropertyMock, TwoOptionsPropertyMock, EnumPropertyMock } from '@shko.online/componentframework-mock';

import { Calendar as Component } from '../Calendar/index';
 
import ReactDOM from 'react-dom';

export type StoryArgs = {
  isDisabled: boolean;
  isVisible: boolean;
  Theme: string;
  BackgroundColor: string;
  Language: string;
  AccessibilityLabel: string;
  SelectedDateValue: Date;
  MinDate: Date;
  MaxDate: Date;
  ShowGoToToday: boolean;
  MonthPickerVisible: boolean;
  DayPickerVisible: boolean;
  HighlightSelectedMonth: boolean;
  HighlightCurrentMonth: boolean;
  ShowSixWeeksByDefault: boolean;
  ShowWeekNumbers: boolean;
  InputEvent: string;
  FirstDayOfWeek: "Sunday" | "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday";
};

const renderGenerator = (
  onInitialize?: (
    mockGenerator: ComponentFrameworkMockGeneratorReact<IInputs, IOutputs>,
    container: HTMLDivElement,
    args: StoryArgs,
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void,
  onUpdate?: (
    mockGenerator: ComponentFrameworkMockGeneratorReact<IInputs, IOutputs>,
    container: HTMLDivElement,
    args: StoryArgs,
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void
) => {
  let container: HTMLDivElement | null;
  let mockGenerator: ComponentFrameworkMockGeneratorReact<IInputs, IOutputs>;

  return function () {
    const [args, updateArgs] = useArgs<StoryArgs>();
    // Fires on unload story
    useEffect(
      () => () => {
        container = null;
        mockGenerator.control.destroy();
      },
      []
    );
    // init logic: 
    if (!container) {
      container = document.createElement('div');
      // initialize ComponentFramework Mock Generator
      mockGenerator = new ComponentFrameworkMockGeneratorReact(
        Component,
        {
            Theme: StringPropertyMock,            
            BackgroundColor: StringPropertyMock,            
            Language: StringPropertyMock,            
            AccessibilityLabel: StringPropertyMock,            
            SelectedDateValue: DateTimePropertyMock,            
            MinDate: DateTimePropertyMock,            
            MaxDate: DateTimePropertyMock,            
            ShowGoToToday: TwoOptionsPropertyMock,            
            MonthPickerVisible: TwoOptionsPropertyMock,            
            DayPickerVisible: TwoOptionsPropertyMock,            
            HighlightSelectedMonth: TwoOptionsPropertyMock,            
            HighlightCurrentMonth: TwoOptionsPropertyMock,            
            ShowSixWeeksByDefault: TwoOptionsPropertyMock,            
            ShowWeekNumbers: TwoOptionsPropertyMock,            
            InputEvent: StringPropertyMock,            
            FirstDayOfWeek: EnumPropertyMock,            
        },
        {
           SelectedDateValue: 'Date', 
        }
      );

      onInitialize?.(mockGenerator, container, args, updateArgs);

      // Initialize default state of the parameters
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context._SetCanvasItems({
            Theme: args.Theme,           
            BackgroundColor: args.BackgroundColor,           
            Language: args.Language,           
            AccessibilityLabel: args.AccessibilityLabel,           
            SelectedDateValue: args.SelectedDateValue,           
            MinDate: args.MinDate,           
            MaxDate: args.MaxDate,           
            ShowGoToToday: args.ShowGoToToday,           
            MonthPickerVisible: args.MonthPickerVisible,           
            DayPickerVisible: args.DayPickerVisible,           
            HighlightSelectedMonth: args.HighlightSelectedMonth,           
            HighlightCurrentMonth: args.HighlightCurrentMonth,           
            ShowSixWeeksByDefault: args.ShowSixWeeksByDefault,           
            ShowWeekNumbers: args.ShowWeekNumbers,           
            InputEvent: args.InputEvent,           
            FirstDayOfWeek: args.FirstDayOfWeek,           
      });

      // Execute the component's init method      
      mockGenerator.ExecuteInit();
    }

    // Update view
    if (mockGenerator) {
      onUpdate?.(mockGenerator, container, args, updateArgs);

      // Update the component's parameters based on storybook's arguments
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.UpdateValues({
            Theme: args.Theme,       
            BackgroundColor: args.BackgroundColor,       
            Language: args.Language,       
            AccessibilityLabel: args.AccessibilityLabel,       
            SelectedDateValue: args.SelectedDateValue,       
            MinDate: args.MinDate,       
            MaxDate: args.MaxDate,       
            ShowGoToToday: args.ShowGoToToday,       
            MonthPickerVisible: args.MonthPickerVisible,       
            DayPickerVisible: args.DayPickerVisible,       
            HighlightSelectedMonth: args.HighlightSelectedMonth,       
            HighlightCurrentMonth: args.HighlightCurrentMonth,       
            ShowSixWeeksByDefault: args.ShowSixWeeksByDefault,       
            ShowWeekNumbers: args.ShowWeekNumbers,       
            InputEvent: args.InputEvent,       
            FirstDayOfWeek: args.FirstDayOfWeek,       
      });
      
      // Execute the component's updateView method 
      ReactDOM.render(mockGenerator.ExecuteUpdateView(), container);     
    }

    // returning the rendered html component after this update
    return container;
  };
};

export default renderGenerator;
"
`;

exports[`Renderer can load template Standard 1`] = `
"/* 
    This file was generated with @shko.online/componentframework-mock-cli (pcf-sb)
    Do not manually update this file as it gets overwritten by pcf-sb based on the Manifest configuration.
*/

import type { Meta, StoryObj } from '@storybook/html';

import type { IInputs, IOutputs } from '../LookupDropdown/generated/ManifestTypes';

import { useArgs, useEffect } from "@storybook/preview-api";

import { ComponentFrameworkMockGenerator, LookupPropertyMock, StringPropertyMock, EnumPropertyMock } from '@shko.online/componentframework-mock';

import { LookupDropdown as Component } from '../LookupDropdown/index';
 

export type StoryArgs = {
  isDisabled: boolean;
  isVisible: boolean;
  lookupfield: ComponentFramework.LookupValue;
  customtext: string;
  customselecttext: string;
  dependentlookupfield: ComponentFramework.LookupValue;
  showRecordImage: "false" | "true";
  showOpenRecordButton: "false" | "true";
};

const renderGenerator = (
  onInitialize?: (
    mockGenerator: ComponentFrameworkMockGenerator<IInputs, IOutputs>,
    container: HTMLDivElement,
    args: StoryArgs,
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void,
  onUpdate?: (
    mockGenerator: ComponentFrameworkMockGenerator<IInputs, IOutputs>,
    container: HTMLDivElement,
    args: StoryArgs,
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void
) => {
  let container: HTMLDivElement | null;
  let mockGenerator: ComponentFrameworkMockGenerator<IInputs, IOutputs>;

  return function () {
    const [args, updateArgs] = useArgs<StoryArgs>();
    // Fires on unload story
    useEffect(
      () => () => {
        container = null;
        mockGenerator.control.destroy();
      },
      []
    );
    // init logic: 
    if (!container) {
      container = document.createElement('div');
      // initialize ComponentFramework Mock Generator
      mockGenerator = new ComponentFrameworkMockGenerator(
        Component,
        {
            lookupfield: LookupPropertyMock,            
            customtext: StringPropertyMock,            
            customselecttext: StringPropertyMock,            
            dependentlookupfield: LookupPropertyMock,            
            showRecordImage: EnumPropertyMock,            
            showOpenRecordButton: EnumPropertyMock,            
        },
        container,
        {
           lookupfield: 'Lookup', 
           dependentlookupfield: 'Lookup', 
        }
      );

      onInitialize?.(mockGenerator, container, args, updateArgs);

      // Initialize default state of the parameters
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context._SetCanvasItems({
            lookupfield: args.lookupfield,           
            customtext: args.customtext,           
            customselecttext: args.customselecttext,           
            dependentlookupfield: args.dependentlookupfield,           
            showRecordImage: args.showRecordImage,           
            showOpenRecordButton: args.showOpenRecordButton,           
      });

      // Execute the component's init method      
      mockGenerator.ExecuteInit();
    }

    // Update view
    if (mockGenerator) {
      onUpdate?.(mockGenerator, container, args, updateArgs);

      // Update the component's parameters based on storybook's arguments
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.UpdateValues({
            lookupfield: args.lookupfield,       
            customtext: args.customtext,       
            customselecttext: args.customselecttext,       
            dependentlookupfield: args.dependentlookupfield,       
            showRecordImage: args.showRecordImage,       
            showOpenRecordButton: args.showOpenRecordButton,       
      });
      
      // Execute the component's updateView method 
      mockGenerator.ExecuteUpdateView();
    }

    // returning the rendered html component after this update
    return container;
  };
};

export default renderGenerator;
"
`;
